packages/dds/merge-tree/src/perspective.ts

```diff
@@ -0,0 +1,239 @@
+	 */
+	public nextSegment(segment: ISegment, forward: boolean = true): ISegment {
+		let next: ISegment | undefined;
+		const action = (seg: ISegment) => {
+			if (
+				isSegmentPresent(
+					seg,
+					this._sequenceNumber,
+					this._clientId,
+					this._refSeq,
+					this._localSequenceNumber,
+				)
+			) {
+				next = seg;
+				return LeafAction.Exit;
+			}
+		};
+		(forward ? forwardExcursion : backwardExcursion)(segment, action);
+		return next ?? (forward ? this._mergeTree.endOfTree : this._mergeTree.startOfTree);
+	}
+
+	/**
+	 * @param segment - The segment to start from.
+	 * @returns the previous segment, or the start of the tree if there is no previous segment.
+	 * @remarks This is a convenient equivalent to calling `nextSegment(segment, false)`.
+	 */
+	public previousSegment(segment: ISegment): ISegment {
+		return this.nextSegment(segment, false);
+	}
+}
+
+/**
+ * @param seg - The segment to check.
+ * @param seq - The latest sequence number to consider.
+ * @param clientId - The ID of the client to consider.
+ * @param refSeq - The reference sequence number,
+ * aka the latest sequence number to consider for other clients' changes.
+ * @param localSeq - The latest local sequence number to consider.
+ * @returns true iff this segment was removed in the given perspective.
+ */
+export function wasRemovedBefore(
+	seg: ISegment,
+	seq: number,
+	clientId: number,
+	refSeq: number,
+	localSeq?: number,
+): boolean {
+	if (
+		seg.removedSeq === UnassignedSequenceNumber &&
+		localSeq !== undefined &&
+		seg.localRemovedSeq !== undefined
+	) {
+		return seg.localRemovedSeq <= localSeq;
+	}
+	if (seg.removedClientIds !== undefined && seg.removedClientIds.includes(clientId)) {
+		return (
+			seg.removedSeq !== undefined &&
+			seg.removedSeq !== UnassignedSequenceNumber &&
+			seg.removedSeq <= seq
+		);
+	}
+	return (
+		seg.removedSeq !== undefined &&
+		seg.removedSeq !== UnassignedSequenceNumber &&
+		seg.removedSeq <= refSeq
+	);
+}
+
+/**
+ * @param seg - The segment to check.
+ * @param seq - The latest sequence number to consider.
+ * @param clientId - The ID of the client to consider.
+ * @param refSeq - The reference sequence number,
+ * aka the latest sequence number to consider for other clients' changes.
+ * @param localSeq - The latest local sequence number to consider.
+ * @returns true iff this segment was moved (aka obliterated) in the given perspective.
+ */
+export function wasMovedBefore(
+	seg: ISegment,
+	seq: number,
+	clientId: number,
+	refSeq: number,
+	localSeq?: number,
+): boolean {
+	if (
+		seg.movedSeq === UnassignedSequenceNumber &&
+		localSeq !== undefined &&
+		seg.localMovedSeq !== undefined
+	) {
+		return seg.localMovedSeq <= localSeq;
+	}
+	if (seg.movedClientIds !== undefined && seg.movedClientIds.includes(clientId)) {
+		return (
+			seg.movedSeq !== undefined &&
+			seg.movedSeq !== UnassignedSequenceNumber &&
+			seg.movedSeq <= seq
```

Comment:
i think this also needs to consider the client id, as a client will always see its own edits

Reply:
It did, but as discussed at standup we don't need to make this distinction between `seq` and `refSeq`.